src/tree/traits.rs

```diff
@@ -0,0 +1,127 @@
+    type ChildIter<'a>: Iterator<Item = NodeId>
+    where
+        Self: 'a;
+
+    /// Get the list of children IDs for the given node
+    fn child_ids(&self, parent_node_id: NodeId) -> Self::ChildIter<'_>;
+
+    /// Get the number of children for the given node
+    fn child_count(&self, parent_node_id: NodeId) -> usize;
+
+    /// Get a specific child of a node, where the index represents the nth child
+    fn get_child_id(&self, parent_node_id: NodeId, child_index: usize) -> NodeId;
+}
+
+/// A marker trait which extends `TraversePartialTree` with the additional guarantee that the child/children methods can be used to recurse
+/// infinitely down the tree. Is required by the `RoundTree` and the `PrintTree` traits.
+pub trait TraverseTree: TraversePartialTree {}
+
+/// Any type that implements [`LayoutPartialTree`] can be laid out using [Taffy's algorithms](crate::compute)
+///
+/// Note that this trait extends [`TraversePartialTree`] (not [`TraverseTree`]). Taffy's algorithm implementations have been designed such that they can be used for a laying out a single
+/// node that only has access to it's immediate children.
+pub trait LayoutPartialTree: TraversePartialTree {
+    /// Get a reference to the [`Style`] for this node.
+    fn get_style(&self, node_id: NodeId) -> &Style;
+
+    /// Set the node's unrounded layout
+    fn set_unrounded_layout(&mut self, node_id: NodeId, layout: &Layout);
+
+    /// Get a mutable reference to the [`Cache`] for this node.
+    fn get_cache_mut(&mut self, node_id: NodeId) -> &mut Cache;
+
+    /// Compute the specified node's size or full layout given the specified constraints
+    fn compute_child_layout(&mut self, node_id: NodeId, inputs: LayoutInput) -> LayoutOutput;
+}
+
+/// Trait used by the `round_layout` method which takes a tree of unrounded float-valued layouts and performs
+/// rounding to snap the values to the pixel grid.
+///
+/// As indicated by it's dependence on `TraverseTree`, it required full recursive access to the tree.
+pub trait RoundTree: TraverseTree {
+    /// Get the node's unrounded layout
+    fn get_unrounded_layout(&self, node_id: NodeId) -> &Layout;
+    /// Get a reference to the node's final layout
+    fn set_final_layout(&mut self, node_id: NodeId, layout: &Layout);
+}
+
+/// Trait used by the `print_tree` method which prints a debug representation
+///
+/// As indicated by it's dependence on `TraverseTree`, it required full recursive access to the tree.
+pub trait PrintTree: TraverseTree {
+    /// Get a debug label for the node (typically the type of node: flexbox, grid, text, image, etc)
+    fn get_debug_label(&self, node_id: NodeId) -> &'static str;
+    /// Get a reference to the node's final layout
+    fn get_final_layout(&self, node_id: NodeId) -> &Layout;
+}
+
+// --- PRIVATE TRAITS
+
+/// A private trait which allows us to add extra convenience methods to types which implement
+/// LayoutTree without making those methods public.
```

Comment:
Can you say more about why we don't want these methods to be public? It's not clear to me on reading this and I suspect it will be challenging to understand years from now.

Reply:
Ah... well it's possible that we do. But that would be *instead* of the `compute_child_layout` method on the `LayoutPartialTree` trait (which would be split into 3: `layout_child`, `measure_child` and `mark_child_as_hidden`). This would be more verbose (dispatch logic would need to be duplicated between `layout_child` and `measure_child` methods), but arguably clearer for implementors. In some cases it also wouldn't be more verbose: the Xilem integration is currently separately implementing the logic within a match statement anyway.

Reply:
The reason these private methods exist is because there used to be 2 separate methods on the trait, and this avoided the need to update all calls in the algorithm implementations.

Reply:
Okay! That's a helpful enough explanation for me for now; feel free to decide how much of this to put into the docs.